Fix: libpacemaker: crm_mon --one-shot can fail during pacemaker shutdown

nrwahl2 opened the pull request:
`crm_mon --one-shot` checks the pacemakerd state before trying to get a CIB connection. If `pacemakerd` is shutting down, it returns `ENOTCONN`. This can cause a resource agent that calls `crm_mon` (for example, `ocf:heartbeat:pgsql`) to fail to stop during shutdown.

This may be a regression introduced by commit 3f342e3. `crm_mon.c:pacemakerd_status()` returns `pcmk_rc_ok` if `pacemakerd` is shutting down, since 49ebe4c and 46d6edd (fixes for CLBZ#5471). 3f342e3 refactored `crm_mon --one-shot` to use library functions, and `pcmk_status.c:pacemakerd_status()` returns `ENOTCONN` if `pacemakerd` is shutting down. As a result, we don't try to connect to the CIB.

Fixes CLBZ#5501

kgaillot commented on lib/pacemaker/pcmk_status.c:
```diff
@@ -122,15 +122,20 @@ pacemakerd_status(pcmk__output_t *out)
+        default:
+            rc = ENOTCONN;
+            break;
+    }
```
EAGAIN would make sense for the init/starting_daemons/wait_for_ping states, where the cluster is coming up and the caller should try again in a bit. ENOTCONN makes sense for shutdown_complete (and invalid, which we shouldn't see).

However one_shot() would still treat EAGAIN as an error. My gut feeling is that we should report the state and return success, since we were able to get a definitive status.

nrwahl2 replied:
I'm working on something in that vein but I don't really like how it's turning out (aside from the added complexity). I think `crm_mon --one-shot` really *should* return an error if it failed to get the status.

The issue at hand is that it's not even trying to get the status if `pacemakerd` is shutting down.

nrwahl2 replied:
If we stick with the simple approach here, we can still change the return code to `EAGAIN` for the states you mentioned. I agree it makes sense for those.

kgaillot replied:
I've got mixed feelings. Getting a response from pacemakerd is successfully getting the status, even if we don't have full CIB/fencing information (yet). I tend to think only failing to communicate with the daemons should be an error exit.

nrwahl2 replied:
See #2905 for an approach that returns success from one-shot mode if we're able to get the `pacemakerd` status. However, as mentioned above, I think it will be fine to return an (appropriate) error code if we get the `pacemakerd` status successfully and still hit an error in `cib_connect()` for example. I don't think it's necessary to return "OK" if we failed to get the cluster status.

nrwahl2 replied:
Hmm yeah, I see your point of view on that. We could just return success from `pcmk__status() no matter what then, if we make it past the `pacemakerd_status()` section. And we could print the `pacemakerd` status from within `pcmk__status()` instead of from within the caller, if we're considering that to be "good enough" as a cluster state result.

kgaillot replied:
I think pacemakerd status should be "good enough" unless we're in running state (because then we *should* have CIB and fencing).

nrwahl2 commented:
Replaced the simple "just enough to unbreak it" fix with a redesign that uses the `pcmk__pacemakerd_status()` function and the `pacemakerd-health` message. It's a lot of commits, but most of them are trivial things that I found along the way. Other than that, it's prep work and then the main change in the last commit.

nrwahl2 commented on lib/pacemaker/pcmk_status.c:
```diff
@@ -289,41 +224,70 @@ pcmk_status(xmlNodePtr *xml)
+                }
+                break;
+            case EREMOTEIO:
+                // @TODO: Why do we try to connect to the fencer and CIB here?
```
I haven't figured out why `EREMOTEIO` is considered okay and we try to connect to the fencer and the CIB after getting that RC. I went ahead and preserved the existing behavior there.

kgaillot replied:
pacemakerd IPC is not proxied to remote nodes, so any attempt to contact it will always fail with EREMOTEIO when run from a remote node. However the fencer and CIB are proxied and may be available.

kgaillot replied:
Sort of, I think the idea is that if the CIB or fencer can be reached, it will immediately be cleared off the screen, but otherwise it'll be visible

nrwahl2 commented on lib/pacemaker/pcmk_cluster_queries.c:
```diff
@@ -195,6 +195,7 @@ pacemakerd_event_cb(pcmk_ipc_api_t *pacemakerd_api,
             break;
 
         default:
+            data->rc = EBADMSG;
```
We could make one or more new Pacemaker return codes for these. `EBADMSG` seemed close enough.

kgaillot replied:
With the pcmk_ipc_api_t model, a single callback is registered for all events, and it's up to the callback to filter out events it's not interested in. So, a different event type wouldn't be a failure, it would just be uninteresting. In other words, the callback isn't deactivated, and when the interesting event happens, the callback will see it.

The other places don't apply because we know what request we made and what reply we should expect. There's no such thing as an extraneous reply, replies match requests.

The other event type that can happen is a notification, which is not related to any request, but is a notification of something potentially interesting that happened on its own -- e.g. in the generic case of any daemon it could be that another client connected, the CIB was replaced, a node was fenced, etc.

nrwahl2 replied:
Okay, I think that makes sense. I was thinking we were done after we hit the callback once. But we keep listening until the timeout expires. Hopefully it's impossible to have a scenario like one client receiving a reply that's intended for another client... i.e., I hope it's guaranteed that a client receives at most one reply, and that that reply is intended for that client.

I guess there's also some mechanism for closing the client and deregistering its callback when it receives a reply. Otherwise with either mainloop or sync dispatch it would keep waiting (and receiving messages) until the timeout.

kgaillot replied:
Right, replies are targeted to the requester, while notifications would be broadcast.

For pcmk_cluster_queries.c, the relevant callbacks call event_done() to close the connection once a final answer is available.

nrwahl2 replied:
BTW just noticed `pcmk_cluster_queries.c:controld_event_reply()` already uses `EBADMSG`. I have no strong feelings; if you prefer `EPROTO`, I need to change it there too.

kgaillot replied:
It's not important -- they should never really happen. Either one conveys the idea.

nrwahl2 commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -647,47 +647,91 @@ health_xml(pcmk__output_t *out, va_list args)
     return out->info(out, "Status of %s: '%s' (last updated %s)",
-                     pcmk__s(sys_from, "unknown node"),
-                     pcmk__s(state, "unknown state"),
+                     pcmk__s(sys_from, "unknown node"), state_s,
```
I kept `"unknown node"` as the `sys_from` default in case it ever can come from a node. As far as I've seen, `sys_from` in this context is always `CRM_SYSTEM_MCP` (`"pacemakerd"`). It might be more accurate to say something like "unknown system", or even to hard-code `CRM_SYSTEM_MCP` if there's no other possibility. (Maybe `"pacemaker-remote"` or similar is a possibility?)

nrwahl2 commented:
Quick note since CI is hung due to infra issues, the rpm target passed on my machine.

kgaillot commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -728,9 +728,10 @@ pacemakerd_health_xml(pcmk__output_t *out, va_list args)
+                                 "sys_from", sys_from,
                                  "state", state_s,
-                                 "last_updated", pcmk__s(last_updated, ""),
+                                 "last_updated", last_updated,
```
Changing the element name isn't really backward compatible since existing scripts won't know that an alternative is possible. Let's keep it as pacemakerd, always hard-coded (I'm not sure why it was sys_from, but that wasn't a good idea -- the whole point of the schema is to have something that scripts can rely on).

nrwahl2 replied:
I was concerned about that. Sounds good, I'll change it to `pacemakerd`.

kgaillot commented on include/pcmki/pcmki_status.h:
```diff
@@ -50,7 +50,7 @@ int pcmk__status(pcmk__output_t *out, cib_t *cib,
                  uint32_t show_opts, const char *only_node,
                  const char *only_rsc, const char *neg_location_prefix,
-                 bool simple_output);
+                 bool simple_output, int timeout_ms);
```
guint makes more sense

nrwahl2 replied:
Yeah you're right. Normally I lean slightly toward "don't use types to mask programmer error", but it's passed as the `message_timeout_ms` (`guint`) argument of `pcmk__pacemakerd_status()`.

kgaillot commented on lib/pacemaker/pcmk_cluster_queries.c:
```diff
@@ -274,9 +274,10 @@ ipc_connect(data_t *data, enum pcmk_ipc_server server, pcmk_ipc_callback_t cb)
+        if (rc != EREMOTEIO) {
+            out->err(out, "error: Could not connect to %s: %s",
+                     pcmk_ipc_name(api, true), pcmk_rc_str(rc));
+        }
```
This means that all other callers would have to check for EREMOTEIO and output the error. Maybe a bool argument for whether EREMOTEIO is ok (probably also need to return pcmk_rc_ok if so)

nrwahl2 replied:
My current thought is to return `EREMOTEIO` from `pcmk_cluster_queries.c:ipc_connect()` if `eremoteio_ok (bool)` is true, but without logging an error. That way if the caller needs to do anything special, it can.

In particular, I'd like `pcmk__status()` to output a `pacemakerd-health` message with state "Running on Pacemaker Remote node". To do that, it needs to know that `ipc_connect()` returned `EREMOTEIO`.

nrwahl2 replied:
We could also just skip the `pacemakerd-health` message altogether in that case. We'd just have less detailed info if we fail to connect to the CIB.

nrwahl2 replied:
...yeah, probably better to just skip `pacemakerd-health` for remote nodes.
1. No clean way to do a short vs. friendly state string for "Running on Pacemaker Remote node"
2. From your comment on another thread, it sounds like "Running on" would be misleading since pacemakerd isn't proxied to remote nodes... and maybe we'd EREMOTEIO even if pacemaker_remoted isn't running, I dunno

kgaillot replied:
Yep this is a good sign you should test the affected commands on remote nodes (not running, running but not connected, connected)

nrwahl2 replied:
In a later PR (not touching it here), I might rethink this... setting `rc = pcmk_rc_ok` within `ipc_connect()` is fine if and only if we also set `data->pcmkd_state = pcmk_pacemakerd_state_running` when we get `EREMOTEIO`. That no longer seems unreasonable. It seems that `EREMOTEIO` is returned if and only if `pacemaker-remoted` is running (regardless of whether it's connected). That's roughly equivalent to `pacemakerd` running.

The current and planned callers treat `EREMOTEIO` the same as `pcmk_rc_ok && pcmk_pacemakerd_state_running`, so we could cut out the middle man by doing it here if there's no objection.

kgaillot replied:
Sounds reasonable

kgaillot commented on lib/common/ipc_pacemakerd.c:
```diff
@@ -62,6 +75,25 @@ pcmk_pacemakerd_api_daemon_state_enum2text(
+        && (state <= pcmk_pacemakerd_state_max)) {
+        return pcmkd_state_friendly_str[state];
+    }
+    return "Invalid pacemakerd state";
```
usually we just use a switch for this sort of thing rather than an array

nrwahl2 replied:
A switch makes more sense and I'll update. I directly modeled this after `pcmk_pacemakerd_api_daemon_state_enum2text` and didn't think about it. The array prevents redundancy there but doesn't do anything for us here.

At some point I want to deprecate those long names and make them aliases for names like `pcmk_pcmkd_state_enum2text()` or something...

kgaillot commented on lib/pacemaker/pcmk_status.c:
```diff
@@ -70,76 +70,38 @@ fencing_connect(void)
-
-    pcmk_register_ipc_callback(pacemakerd_api, pacemakerd_event_cb, (void *) &state);
-
-    rc = pcmk_connect_ipc(pacemakerd_api, pcmk_ipc_dispatch_sync);
```
I'm not sure how we should handle the dispatch type. I believe it was sync here because pcmk_status() might be called without a mainloop. pcmk__pacemakerd_status() runs its own mainloop, so that might handle that situation, but I'm not sure that's a good idea if the caller already has one, or how we might integrate with an existing mainloop (which shouldn't end once we get a reply). Hmmm ...

nrwahl2 replied:
Can there only be one at a time globally? Otherwise I don't see the harm in having this second mainloop temporarily that's only reachable from the local `data` object. I'm not going to pretend that I'm well-acquainted with how either GLib's or Pacemaker's main loop code works though, including the idea of a GMainContext.

We could always have a bool or an enum that says which type of dispatch to use, or use sync automatically if the timeout is 0. It would complicate the interface and implementation a little bit, but that might be preferable to reasoning about multiple mainloops if we don't feel confident.

kgaillot replied:
There can be more than one mainloop, but one will be the default, and pacemaker attaches various things to the default mainloop. That's just another gotcha to worry about -- we'd have to make sure this mainloop can't be the default, and make sure whatever we do with it doesn't assume the default mainloop is desired.

Using sync with timeout==0 sounds reasonable.

nrwahl2 replied:
Can we do that in `ipc_connect()` (set to sync if timeout == 0), or do we need to do it elsewhere to avoid affecting the public callers' behavior?

Another less pretty option is to pass a dispatch type to `pcmk__pacemakerd_status()`, `pcmk__controller_status()`, `pcmk__designated_controller()`, and `ipc_connect()`. That way for example `pcmk__status()` can choose sync if timeout == 0, but the public `pcmk_pacemakerd_status()` function will still get pcmk_ipc_dispatch_main with default timeout 30000.

kgaillot replied:
We can keep public calls the same by passing the default timeout explicitly

nrwahl2 replied:
Oh yeah good point

kgaillot replied:
Though it might make sense for public calls to use sync. Not a whole lot of point in running a mainloop that essentially emulates a sync call

nrwahl2 replied:
Yeah, only difference I'm aware of is the 30s vs 5s timeout

kgaillot reviewed (commented):
Whew, this got big ...

nrwahl2 commented on lib/pacemaker/pcmk_cluster_queries.c:
```diff
@@ -36,6 +36,7 @@ typedef struct {
     int rc;
     guint message_timer_id;
     guint message_timeout_ms;
+    void *private;
```
Definitely can. The idea was that the `private` member would be simple to extend later to use for something else in other callbacks, whereas that particular enum likely wouldn't be. For example, the same `data_t` struct type is passed to `controller_status_event_cb()` and `designated_controller_event_cb()`.

Since this is internal as you mentioned, it also wouldn't add multiple new fields as needed, and either use them or ignore them as appropriate.

nrwahl2 commented:
Updated to address review. I think the below bullet points cover all the changes...
* Use switch instead of array for `enum2friendly()`
* Use `guint` for `timeout_ms` argument to `pcmk__status()`
* Added commit on top to change `sys_from` default to `"unknown subsystem"`
* Don't return an error for unexpected event type.
* Use an `enum pcmk_pacemakerd_state` in `pcmk_cluster_queries.c:data_t` instead of `void *private`
* Use hardcoded `"pacemakerd"` instead of hardcoded `"pacemakerd-health"` for XML element name, for backward compatibility
* Use `pcmk_ipc_dispatch_sync` instead of `pcmk_ipc_dispatch_main` if `message_timeout_ms == 0` in `pcmk_cluster_queries.c`. This lets us avoid possible complications if the caller has its own mainloop.
* Add an `eremoteio_ok` boolean argument to `pcmk_cluster_queries.c:ipc_connect`. If true, don't log an error message for `EREMOTEIO`. We still return `EREMOTEIO` because the caller may need to handle it specially. In particular, `state` will be invalid, so we can't rely on `pcmk_rc_ok && <a good state>` in our switch statement in `pcmk__status()`. We need to treat `EREMOTEIO` as automatically okay in `pcmk__status()`.

kgaillot commented on xml/api/pacemakerd-health-2.25.rng:
```diff
@@ -0,0 +1,28 @@
+            <element name="pacemakerd">
+                <attribute name="state"> <text /> </attribute>
+                <attribute name="last_updated"> <text /> </attribute>
+            </element>
```
The second pacemakerd element isn't necessary. The first covers this case as well. (Going from mandatory to optional is backward compatible, as older output will still validate.)

nrwahl2 replied:
Yep, forgot to update this after s/pacemakerd-health/pacemakerd/

kgaillot commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -728,9 +728,10 @@ pacemakerd_health_xml(pcmk__output_t *out, va_list args)
     }
 
-    pcmk__output_create_xml_node(out, pcmk__s(sys_from, "<null>"),
+    pcmk__output_create_xml_node(out, "pacemakerd",
```
Update commit message ("Fix pacemakerd-health XML output")

kgaillot reviewed (commented):
Just about there

nrwahl2 commented:
Rebased on current main and updated to address review

nrwahl2 commented on include/pacemaker.h:
```diff
@@ -111,7 +111,13 @@ void pcmk_free_injections(pcmk_injections_t *injections);
+ *                                    If positive, \p pcmk_ipc_dispatch_main
+ *                                    will be used, and a new mainloop will be
+ *                                    created for this purpose (freed before
+ *                                    return).
```
I wonder if it would be better to replace the mainloop stuff in `pcmk_cluster_queries.c` with `pcmk_ipc_dispatch_poll` when we need a timeout. That way we never have to worry about mainloop conflicts.

If so, it'll go in a separate PR.

kgaillot replied:
Makes sense, as long as anything `crm_mon` needs still works

nrwahl2 replied:
Ack. `crm_mon` already uses polling (timeout == reconnect_ms / 2) instead of sync or mainloop, for its console and daemonized modes

gao-yan commented on lib/pacemaker/pcmk_cluster_queries.c:
```diff
@@ -334,16 +345,26 @@ pcmk__designated_controller(pcmk__output_t *out, guint message_timeout_ms)
+    pcmk_ipc_api_t *controld_api = NULL;
+
+    if (message_timeout_ms == 0) {
+        dispatch_type = pcmk_ipc_dispatch_sync;
```
AFAICT this also changed the behavior of "crmadmin -D" under certain circumstance.

Now upon cluster startup, if "crmadmin -D" is executed before a DC is elected yet, it will be hanging forever .

AFAICS the lookup involves a ping request to the DC. But given that there's not a DC to actually handle the request at the moment,  crmadmin most probably won't receive any reply under the situation...

Previously with an implied timeout 30s, the command would eventually return with an error though rather than hanging forever.

Or course first it's a question what should be the best as the default for such a case.

nrwahl2 replied:
The idea was that synchronous requests are supposed to time out after 5 seconds...

> @nrwahl2 , do you want to look into that?

"want" is such a loaded word ;)

nrwahl2 replied:
https://projects.clusterlabs.org/T735

gao-yan replied:
> https://projects.clusterlabs.org/T735

Good thinking over there. 

My two cents. Unless I'm missing anything, it sounds good enough to me for "crmadmin -D" to have the same understanding as the local controller in regard of who is the DC. If our controller doesn't think there's a DC yet, probably so be it and just tell crmadmin the same? 

Basically it probably doesn't really need to cost a ping request to DC to know who is the DC. It would be more meaningful if we'd like to achieve the following with a single command/request:

`crmadmin -S $(crmadmin -D -q)`

gao-yan replied:
> Now upon cluster startup, if "crmadmin -D" is executed before a DC is elected yet, it will be hanging forever .
> 
> AFAICS the lookup involves a ping request to the DC. But given that there's not a DC to actually handle the request at the moment, crmadmin most probably won't receive any reply under the situation...

Another exposed issue is, without anything to read from the IPC at the moment, the looping in pcmk__send_ipc_request() becomes extremely busy since crm_ipc_read() keeps getting -EAGAIN:

https://github.com/ClusterLabs/pacemaker/blob/main/lib/common/ipc_client.c#L703-L706

Maybe a sleep()  or pcmk_poll_ipc() before retry would make more sense?

nrwahl2 replied:
`sleep()` would still be synchronous, and I think `pcmk_poll_ipc()` would be just as busy -- it's just shifting the work to the `poll()` system call. It blocks on `poll()` until either we receive something or the timeout we pass expires. I mentioned in T735 that at a glance, it looks like `pcmk_ipc_dispatch_sync` might behave the same as `pcmk_ipc_dispatch_poll` with no timeout.

In theory we could add an asynchronous sleep via a mainloop -- but at that point, we should use `pcmk_ipc_dispatch_main` if a mainloop is available.

gao-yan replied:
> `sleep()` would still be synchronous (although it would be less work, at the expense of a delay in finding a message)
> 
> I think `pcmk_poll_ipc()` would be just as busy -- it's just shifting the work to the `poll()` system call. It blocks on `poll()` until either we receive something or the timeout we pass expires.

Blocking on one poll() call is better than constantly iterating it. Since crm_ipc_read() calls qb_ipc_us_ready() -> poll() with a timeout 0,  continuously iterating it without an interval due to EAGAIN significantly raises CPU load.

> I mentioned in T735 that at a glance, it looks like `pcmk_ipc_dispatch_sync` might behave the same as `pcmk_ipc_dispatch_poll` with no timeout.

Indeed.

kgaillot replied:
Hmm, the original thinking was that EAGAIN would be a brief condition (often one or two tries). It's definitely a problem that timeouts would have no effect if it's stuck in that loop. It's especially bad if a daemon is waiting on another daemon.

I'm thinking we could poll with a 5s timeout.

An async wait via mainloop might be better, especially for daemons, but it's impractical (every send would need a callback for when it completes), and we'd still need the timeout for non-mainloop sends.

gao-yan replied:
Is there anything about this yet by any chance? Otherwise should I try doing something about this? :-)

kgaillot replied:
@gao-yan I don't think anything's been started yet

gao-yan replied:
Okay, thanks for the update. I'll take a look then.
